### PR TITLE
RSDK-2498: adding intel up4000

### PR DESCRIPTION
**Pull request by susmitaSanyal**
Adding new board component; Intel up4000. 
Tested genericlinux changes on pi. 

TODO: 

- [ ] Test on jetson and beagle

- [ ] Test HW PWM

### CODE REVIEW

File: components/board/intel/board.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
+// Package intel implements a Intel based board.
+package intel
+
```
Comment (randhid) [inline components/board/intel/board.go:3]: [minor] add tested models, datasheets or references, etc.

File: components/board/intel/board.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import (
+	"github.com/edaniels/golog"
+	"github.com/pkg/errors"
+	"periph.io/x/host/v3"
+
+	"go.viam.com/rdk/components/board/genericlinux"
+)
+
+const modelName = "intel"
```
Comment (randhid) [inline components/board/intel/board.go:12]: [major] We don't have a scop with a model name yet, so let's not register this fully just yet, as that is a breaking change.

File: components/board/genericlinux/data.go [hunk 3]
```diff
@@ -96,16 +97,19 @@ func GetGPIOBoardMappings(modelName string, boardInfoMappings map[string]BoardIn
...
-			return nil, noBoardError(modelName)
-		}
-		return nil, err
+	var path string
+	const (
+		compatiblePath    = "/proc/device-tree/compatible"
+		compatiblePathDmi = "/sys/devices/virtual/dmi/id/board_name"
+	)
+
+	if modelName == "intel" {
```
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/board/genericlinux/data.go:106]: Agreed. Does RDK store architecture information? if not I can create a function.

**Comment by penguinland** [comment]
> Tested genericlinux changes on pi.

Rand is right: the pi is the only board that _doesn't_ use genericlinux, and we should test this on a BeagleBone and/or Jetson.

File: components/board/genericlinux/data.go [hunk 4]
```diff
@@ -96,16 +97,22 @@ func GetGPIOBoardMappings(modelName string, boardInfoMappings map[string]BoardIn
...
+		path = compatiblePathDmi
+	} else {
+		path = compatiblePath
+	}
 
-	compatiblesRd, err := os.ReadFile(compatiblePath)
+	compatibles, err := newStringSetFromFile(modelName, path)
```
Comment (penguinland) [inline components/board/genericlinux/data.go:112]: I guess this works, but the helper function I would have created is `getCompatibles`. Something like this:

```
func getCompatibles(modelName string) []string, err {
    const (
        compatiblePathDt = "/proc/device-tree/compatible"
        compatiblePathDmi = "/sys/devices/virtual/dmi/id/board_name"
    )
    if compatiblePathDmi is a valid file:
        return a length-1 StringSet containing its contents

    read compatiblePathDt, return its contents split on null bytes
}
```

Making the `path` variable have two different purposes and passing that around seems like extra mental overhead for anyone working on this code, and then trying to write a single implementation of `newStringSetFromFile` that supports both device-tree and DMI boards seems hard.

File: components/board/genericlinux/data.go [hunk 5]
```diff
@@ -96,16 +97,22 @@ func GetGPIOBoardMappings(modelName string, boardInfoMappings map[string]BoardIn
 // getCompatiblePinDefs returns a list of pin definitions, from the first BoardInformation struct
 // that appears compatible with the machine we're running on.
 func getCompatiblePinDefs(modelName string, boardInfoMappings map[string]BoardInformation) ([]PinDefinition, error) {
-	const compatiblePath = "/proc/device-tree/compatible"
+	var path string
+	const (
+		compatiblePath    = "/proc/device-tree/compatible"
```
Comment (penguinland) [inline components/board/genericlinux/data.go:102]: [minor] This sounds like it's for all boards, but you've implemented the rest as though this is only for device-tree boards. I'd put a `Dt` suffix on the name, similar to the `Dmi` suffix on the next line.

File: components/board/intel/board.go (hunk 2 above)
Comment (penguinland) [reply components/board/intel/board.go:12]: Agreed: there are likely to be many Intel boards, but this only supports one of them for now. Maybe this should be "up4000", or maybe the plan is to expand this to support lots of boards by having lots of definitions in data.go? but I'd like a plan for how this works before we commit to such a broad model name.

**Review by penguinland — CHANGES_REQUESTED** [review]
The proof that this works should be successfully controlling GPIO pins on the Jetson and Up4000. Doing anything on the rpi is unrelated.

File: components/board/genericlinux/data.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (randhid) [reply components/board/genericlinux/data.go:106]: https://pkg.go.dev/runtime#pkg-constants

File: components/board/intel/board.go (hunk 2 above)
Comment (susmitaSanyal) [reply components/board/intel/board.go:12]: Will update this when we decide on the name in scope doc. Till then, I'll keep "intel" as a placeholder.

File: components/board/register/register.go [hunk 6]
```diff
@@ -7,6 +7,7 @@ import (
 	_ "go.viam.com/rdk/components/board/beaglebone"
 	_ "go.viam.com/rdk/components/board/fake"
 	_ "go.viam.com/rdk/components/board/hat/pca9685"
+	_ "go.viam.com/rdk/components/board/intel"
```
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/board/register/register.go:10]: - [ ] remove component from register.

File: components/board/genericlinux/data.go [hunk 7]
```diff
@@ -120,9 +127,28 @@ func getCompatiblePinDefs(modelName string, boardInfoMappings map[string]BoardIn
 	if pinDefs == nil {
 		return nil, noBoardError(modelName)
 	}
+
```
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/board/genericlinux/data.go:129]: a lot of these extra line was added and removed while I was printing out debug statements. Please ignore those, I will address them as they show up.

File: components/board/register/register.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (randhid) [reply components/board/register/register.go:10]: hold off on this for now - I'm going to see if we can get a name today to unblock you.

[the remark this answers is not in the document]
Comment (randhid) [reply components/board/register/register.go:10]:
```suggestion
	_ "go.viam.com/rdk/components/board/upboard"
```

File: components/board/intel/board.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
```
Comment (randhid) [inline components/board/intel/board.go]:
```suggestion
const modelName = "upboard"
```

File: components/board/intel/board.go [hunk 9]
```diff
@@ -0,0 +1,32 @@
+// Package intel implements an Intel based board.
```
Comment (randhid) [inline components/board/intel/board.go]:
```suggestion
// Package upboard implements an Intel based board.
```

File: components/board/intel/board.go [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+package intel
```
Comment (randhid) [inline components/board/intel/board.go]:
```suggestion
package upboard
```

File: components/board/upboard/data.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
+package upboard
+
+import "go.viam.com/rdk/components/board/genericlinux"
+
+const upboard = "up_4000"
```
Comment (randhid) [inline components/board/upboard/data.go:5]:
```suggestion
const upboard = "up4000"
```

**Pull request closed without merging**